cpp/src/sampling/negative_sampling_impl.cuh

```diff
@@ -0,0 +1,406 @@
+          bool multi_gpu>
+rmm::device_uvector<vertex_t> create_local_samples(
+  raft::handle_t const& handle,
+  raft::random::RngState& rng_state,
+  graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu> const& graph_view,
+  std::optional<rmm::device_uvector<weight_t>> const& normalized_biases,
+  std::optional<rmm::device_uvector<weight_t>> const& gpu_biases,
+  size_t samples_in_this_batch)
+{
+  rmm::device_uvector<vertex_t> samples(0, handle.get_stream());
+
+  if (normalized_biases) {
+    size_t samples_to_generate{samples_in_this_batch};
+    std::vector<size_t> sample_count_from_each_gpu;
+
+    rmm::device_uvector<size_t> position(0, handle.get_stream());
+
+    if constexpr (multi_gpu) {
+      // Determine how many vertices are generated on each GPU
+      auto const comm_size = handle.get_comms().get_size();
+      auto const rank      = handle.get_comms().get_rank();
+      auto& major_comm     = handle.get_subcomm(cugraph::partition_manager::major_comm_name());
+      auto const major_comm_size = major_comm.get_size();
+      auto& minor_comm = handle.get_subcomm(cugraph::partition_manager::minor_comm_name());
+      auto const minor_comm_size = minor_comm.get_size();
+
+      sample_count_from_each_gpu.resize(comm_size);
+
+      rmm::device_uvector<size_t> gpu_counts(comm_size, handle.get_stream());
+      position.resize(samples_in_this_batch, handle.get_stream());
+
+      thrust::fill(handle.get_thrust_policy(), gpu_counts.begin(), gpu_counts.end(), size_t{0});
+      thrust::sequence(handle.get_thrust_policy(), position.begin(), position.end());
+
+      rmm::device_uvector<weight_t> random_values(samples_in_this_batch, handle.get_stream());
+      detail::uniform_random_fill(handle.get_stream(),
+                                  random_values.data(),
+                                  random_values.size(),
+                                  weight_t{0},
+                                  weight_t{1},
+                                  rng_state);
+
+      thrust::sort(handle.get_thrust_policy(),
+                   thrust::make_zip_iterator(random_values.begin(), position.begin()),
+                   thrust::make_zip_iterator(random_values.end(), position.end()));
+
+      thrust::upper_bound(handle.get_thrust_policy(),
+                          random_values.begin(),
+                          random_values.end(),
+                          gpu_biases->begin(),
+                          gpu_biases->end(),
+                          gpu_counts.begin());
+
+      thrust::adjacent_difference(
+        handle.get_thrust_policy(), gpu_counts.begin(), gpu_counts.end(), gpu_counts.begin());
+
+      // all_gpu_counts[i][j] will be how many vertices need to be generated on GPU j to be sent to
+      // GPU i
+      auto all_gpu_counts = cugraph::device_allgatherv(
+        handle,
+        handle.get_comms(),
+        raft::device_span<size_t const>{gpu_counts.data(), gpu_counts.size()});
```

Comment:
You need to just collect sample counts for comm_size GPUs (no need to collect for the entire comm_size * comm_size matrix). Use device_multicast_sendrecv to send to comm_size GPUs and receive from comm_size GPUs.

Reply:
Then, the code below will also become simpler.